indigo_api/data_migrations/__init__.py

```diff
@@ -94,3 +113,239 @@ def update_namespace(self, xml):
+            for node in doc.root.xpath(f"//a:{element}", namespaces=nsmap):
+                old_id = node.get("id")
+                # note: this assumes that all subsections and items have <num>s, which _should_ be true
+                num = node.num.text
+                num = self.clean_number(num)
+                prefix = self.get_parent_id(node)
+                new_id = f"{prefix}.{replacement}{num}"
+                self.safe_update(node, mappings, old_id, new_id)
+
+        # generate correct term ids, based on the parent id
+        counter = Counter()
+        for node in doc.root.xpath('//a:term', namespaces=nsmap):
+            old_id = node.get('id')
+            prefix = self.get_parent_id(node)
+            # the num for this term depends on the number of preceding terms with the same prefix
+            counter[prefix] += 1
+            new_id = f'{prefix}__term_{counter[prefix]}'
+            self.safe_update(node, mappings, old_id, new_id)
+
+        # finally
+        # "." separators
+        for node in doc.root.xpath("//a:*[@id]", namespaces=nsmap):
+            old_id = node.get("id")
+            if "." in old_id:
+                new_id = old_id.replace(".", "__")
+                node.set("id", new_id)
+                mappings.update({old_id: new_id})
+
+        # replace all `id`s with `eId`s
+        for node in doc.root.xpath("//a:*[@id]", namespaces=nsmap):
+            node.set("eId", node.get("id"))
+            del node.attrib["id"]
+
+    def clean_number(self, num):
+        # e.g. 1.3.2. / (3) / (dA) / (12 bis)
+        return num.strip("().").replace(" ", "").replace(".", "-")
+
+    def get_parent_id(self, node):
+        parent = node.getparent()
+        parent_id = parent.get("id")
+        return parent_id if parent_id else self.get_parent_id(parent)
+
+
+class HrefMigration(AKNMigration):
+    """ Update all internal cross-references in a document's XML.
+        Note: this will only update references if they still existed in the document at the time of running the previous migrations.
+        If the section referred to was removed from the document for some reason in the interim, its id won't have been updated and the href therefore won't be in `mappings`.
+        The reference would point at nothing in any case, but the nothing it points to would be an old-style id.
+    """
+    def migrate_act(self, doc, mappings):
+        for node in doc.root.xpath("//a:ref[starts-with(@href, '#')]", namespaces={"a": doc.namespace}):
+            ref = node.get("href").lstrip("#")
+            new_ref = self.traverse_mappings(mappings, ref)
+            node.set("href", f"#{new_ref}")
+
+
+class AnnotationsMigration(AKNMigration):
+    """ Update all the annotation anchors on a Document
+    """
+    def migrate_act(self, doc, mappings):
+        for annotation in doc.annotations.all():
+            annotation.anchor_id = self.traverse_mappings(mappings, annotation.anchor_id)
```

Comment:
For a comment in a schedule, the old id will have a format such as `schedule/schedule.paragraph0`, but the new id will not have that schedule prefix.